### ISSUE TITLE
implement better threading model vs crash (uncaught exception thrown) situations

### ISSUE THREAD

**Issue by FyiurAmron** [libgdx/libgdx#3240]
In current implementation, if the main app thread (e.g. the `render()` loop) throws an uncaught exception, it kills the main thread properly, but all the async threads (AssetManager's AsyncExecutor instances, Net's ExecutorService, Timers for input/gesture listener etc) are left hanging, with no real way to shutdown them. Sometimes it's desirable to let the exception propagate (e.g. Google Play Services report those nicely) - in theory it would be possible to catch this exception, kill all async threads and then rethrow the exception, closing the whole app. Currently it's impossible without heavy modifications to those (threading) classes.

Note that e.g. calling `assetManager.dispose()` to kill the AsyncExecutor is often impossible - because the OpenGL context might have been lost already (e.g. by some crash in it), disposing OpenGL-managed resources will crash with "OpenGL context lost" exception, leaving the executor thread running. OTOH, for current Net class accessing the executor is completely impossible (private field, no shutdown method, see related link).

Current behaviour might produce long "freezes" to the user after the crash (most visible on desktop) due to those executor threads waiting for timeout (about 30s to 1 minute in my experience). After manually extending all of the thread-creating classes with thread shutdown routines and calling them on exceptional situations, the freeze disappears.

My proposition would be to either implement a automatic libGDX thread "manager", simply keeping track of all libGDX-related threads running & allowing easy scheduled forced termination of them - or to implement a manual scheme (e.g. `GdxThreading` interface with method `shutdownThreads()` clearly documented as "debug/crash usage only" or something similar), allowing the programmer to easily schedule those by himself. This interface would be also a visible marker of the fact that the class *might* create new threads.

NB trying to `stop()`, `interrupt()` or otherwise kill those threads either doesn't work or produces even more exceptions along the way.

tl;dr IMO the libGDX's threads cleanup should be either done by libGDX, or there should be an API facility allowing the programmer to do so.

related: https://github.com/libgdx/libgdx/issues/3226

**Comment by Antag99** [comment]
[Daemon Threads](http://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html#setDaemon%28boolean%29) do not prevent an application from exiting. It looks like `AsyncExecutor` actually uses them; so an `AssetManager` should not prevent the application from exiting once the main thread has been stopped.

To fix #3226, the `ExecutorService` in `NetJavaImpl` should be constructed like it's done in `AsyncExecutor`, so the threads it's creating are daemon. It may also be worth checking if any other part of libgdx uses `ExecutorService` without daemon threads. Actually, most uses of `ExecutorService` could probably be replaced with `AsyncExecutor`.

**Comment by Antag99** [comment]
See [this answer on StackOverflow](http://stackoverflow.com/a/13883412/4517740).

**Comment by badlogic** [comment]
Jupp, daemon threads all the way.

**Comment by FyiurAmron** [comment]
@Antag99 I understand you actually mean http://stackoverflow.com/a/29453160/719662 <- this answer? Also, I too think that daemon threads may be the best solution 

As to the AssetManager-s - it's kinda possible that they aren't what causes that hang, and I guess it's possible their threads are indeed daemon (didn't have enough time to check it all). What I am 100% sure that a) killing their threads (even if they are daemons) is currently impossible in case of some emergency situation (due to above `dispose()` problem), b) in my case, even when fixing `NetJavaImpl` behaviour by hand, there were still 6 non-system threads hanging - 4 from asset managers, 1 LWJGL Timer & 1 Timer (actually used by gesture detector for timing input events), and at least one of those prevented the app from closing. In case AM has daemons (highty probable), IMO it's one of the timers that ain't "daemoning" the way they should.

**Comment by Antag99** [comment]
> @Antag99 I understand you actually mean http://stackoverflow.com/a/29453160/719662 <- this answer? Also, I too think that daemon threads may be the best solution

I don't mean that answer. I mean the answer I linked to.

> killing their threads (even if they are daemons) is currently impossible in case of some emergency situation (due to above dispose() problem)

There is no need to kill their threads. An application terminates after all non-daemon threads have been finished.

> in my case, even when fixing NetJavaImpl behaviour by hand, there were still 6 non-system threads hanging - 4 from asset managers, 1 LWJGL Timer & 1 Timer (actually used by gesture detector for timing input events), and at least one of those prevented the app from closing. In case AM has daemons (highty probable), IMO it's one of the timers that ain't "daemoning" the way they should.

I don't know if the timers use daemon threads. If they don't, it's still possible to kill the application using `System.exit()` - there is no need to perform any cleanup.

**Comment by Antag99** [comment]
It seems like libgdx's `Timer` class uses non-daemon threads. Could you provide more details regarding the LWJGL timer? To me it seems like it does not use threading at all.

**Comment by FyiurAmron** [comment]
a) what's the exact benefit of using daemon threads vs using ThreadPoolExecutor? I'm not saying there ain't any, but frankly I don't see any obvious one here,
b) assume you have a running app, and AM got "broke" due to invalid (indisposable) assets due to some obscure problem. The app identified the situation, let's say it can easily fix that by just dumping the old AM and creating a new one, which will work OK. Still, you can't call `dispose()` on the old one to get rid of its daemon, since it will just try to dispose the indisposable assets, throw an exception and keep the daemons running. Should I just say "yeah, I have one useless thread, without any possibility of killing it, no biggy" or "just kill the app, the user won't notice"?
API should always allow shuting down spawned threads without forcing the programmer to kill the app or do low-level thread management on the VM. There *is* a need for a fool-proof method to be able to shut down the threads as needed. If AM `dispose()` would allow that (e.g. putting the thread disposal in `finally` block in the `dispose()` method - I'd see no problem with that.
c) using `System.exit()` is the worst possible solution - especially on Android. Ever used AdMob with apps using `System.exit()`? AFAIR, with some versions, you won't get ads anymore (was a case for me). Also, http://stackoverflow.com/a/2632649/719662 . The list of problems with using `System.exit()` is quite long, even when intended as an emergency measure.

As to the `LWJGL Timer` thread - https://github.com/LWJGL/lwjgl/blob/master/src/java/org/lwjgl/opengl/Sync.java#L113

**Comment by Antag99** [comment]
> a) what's the exact benefit of using daemon threads vs using ThreadPoolExecutor? I'm not saying there ain't any, but frankly I don't see any obvious one here,

There is no reason to use `ThreadPoolExecutor`.

> b) assume you have a running app, and AM got "broke" due to invalid (indisposable) assets due to some obscure problem. The app identified the situation, let's say it can easily fix that by just dumping the old AM and creating a new one, which will work OK. Still, you can't call dispose() on the old one to get rid of its daemon, since it will just try to dispose the indisposable assets, throw an exception and keep the daemons running. Should I just say "yeah, I have one useless thread, without any possibility of killing it, no biggy" or "just kill the app, the user won't notice"?

The obscure problem should bring down the app, so the developer can fix it. This is why exceptions exist in the first place. Also, sorry regarding `System.exit`; this would swallow the exception. Daemons do not interfere with the termination of an application. **There is no need to kill them.**

> As to the LWJGL Timer thread - https://github.com/LWJGL/lwjgl/blob/master/src/java/org/lwjgl/opengl/Sync.java#L113

It's a daemon. No problem there.

> API should always allow shuting down spawned threads without forcing the programmer to kill the app or do low-level thread management on the VM. There is a need for a fool-proof method to be able to shut down the threads as needed. If AM dispose() would allow that (e.g. putting try around close() and putting https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/assets/AssetManager.java#L603 line in finally block), I'd be more than happy.

The developer should not need to care about terminating threads.

> c) using System.exit() is the worst possible solution - especially on Android. Ever used AdMob with apps using System.exit()? AFAIR, with some versions, you won't get ads anymore (was a case for me). Also, http://stackoverflow.com/a/2632649/719662 . The list of problems with using System.exit() is quite long, even when intended as an emergency measure.

Nevermind that, we're not goint to use `System.exit`. I meant it would be used when an unchecked exception was thrown, didn't think too much there.

I think we should solve this issue by using daemon threads in `NetJavaImpl` and `Timer`.

**Comment by FyiurAmron** [comment]
>The obscure problem should bring down the app, so the developer can fix it. This is why exceptions exist in the first place.

Common UX rules say that if there is a way of dealing with exception that will not disrupt the user's experience, it should be used. Exceptions should be thrown to *signal* an exceptional case and report the error to the developer - not to *crash an app*. If the application can deal with the exception - it should. OTOH, API should allow the developer to deal with the exceptions within reason.

I've had a couple of dozens of user's feedback to *suppress* the exceptions that ain't blocking the execution from doing so. Many of the exceptions are platform-dependent and/or generated by dependent libs and/or natives - and essentially both impossible to predict and to fix. 

I'm not saying (quote me where I said it) that the daemons interfere with exiting. I'm saying that *daemon threads can interfere with execution*, by hogging mem & CPU. On slower devices, this can be visible.

Also, the final decision as to how to deal with exceptional cases should be developer's, not API designer's, one.

>The developer should not need to care about terminating threads.

I agree, that's exactly why I proposed the regular `dispose()` to do that automatically. Currently it doesn't, the simplest test case being shown above.

**Comment by Antag99** [comment]
> Common UX rules say that if there is a way of dealing with exception that will not disrupt the user's experience, it should be used. Exceptions should be thrown to signal an exceptional case and report the error to the developer - not to crash an app. If the application can deal with the exception - it should. OTOH, API should allow the developer to deal with the exceptions within reason.

The case you mentioned is not an error that should be handled. If some assets `dispose()` is raising an exception, something is clearly wrong.

> I've had a couple of dozens of user's feedback to suppress the exceptions that ain't blocking the execution from doing so. Many of the exceptions are platform-dependent and/or generated by dependent libs and/or natives - and essentially both impossible to predict and to fix.

That sounds *bad*. There might be cases where suppressing an exception is *ok*, but it's always a hack.

> Also, the final decision as to how to deal with exceptional cases should be developer's, not API designer's, one.

There is no *need* of dealing with exceptional cases; the application should just crash badly and give a nice error report.

> I agree, that's exactly why I proposed the regular dispose() to do that automatically. Currently it doesn't, the simplest test case being shown above.

It does.

**Comment by Antag99** [comment]
@badlogic I opened a pull request for this: #3257.
